### PR TITLE
feat: implement basic threading support for SGX

### PR DESCRIPTION
**Pull request by haraldh**
This patch series implements basic threading support for the SGX backend.

Currently only two additional threads can be started (and not restarted), because the TCS, SSA and stacks are statically allocated in the linker file. This will be turned into dynamic allocation in a followup PR.

Related https://github.com/enarx/enarx/issues/2160

### CODE REVIEW

File: src/backend/sgx/thread.rs
Comment (rvolosatovs) [inline src/backend/sgx/thread.rs:46]: What's happening here? :thinking:

**Review by rvolosatovs — COMMENTED** [review]
Overall, implementation looks sane, I can't say much about the internals of this (yet)

I will have to dive deeper into this once I get home and have access to a bigger screen

File: src/backend/sgx/thread.rs
Comment (haraldh) [reply src/backend/sgx/thread.rs:46]: read the commit message for this chunk

Comment (bstrie) [reply src/backend/sgx/thread.rs:46]: Does this comment represent a FIXME? If so, it should mention that explicitly.

File: crates/sallyport/src/guest/call/enarxcall/types/result.rs
Comment (bstrie) [inline crates/sallyport/src/guest/call/enarxcall/types/result.rs:51]: `errno @ ERRNO_START.. =>` should suffice.

File: src/backend/sgx/thread.rs
Comment (bstrie) [inline src/backend/sgx/thread.rs:120]: Would `u64::MAX` be better?

File: crates/shim-sgx/src/thread.rs
Comment (haraldh) [reply crates/shim-sgx/src/thread.rs:99]: Yes, this is correct. It loads the instruction pointers and the stack and everything and will never return from this function.

File: crates/shim-sgx/src/handler/mod.rs
Comment (haraldh) [reply crates/shim-sgx/src/handler/mod.rs:239]: Hmm, if we ever pimp up `attacked()` then this is the right action here.

File: src/backend/sgx/thread.rs
Comment (haraldh) [reply src/backend/sgx/thread.rs:120]: true

File: crates/sallyport/src/guest/call/enarxcall/types/result.rs
Comment (haraldh) [reply crates/sallyport/src/guest/call/enarxcall/types/result.rs:51]: copy & pasted from existing sallyport code, but you are right.

Comment (rvolosatovs) [reply crates/sallyport/src/guest/call/enarxcall/types/result.rs:51]: I believe unbounded ranges were unstable when most of sallyport was written.
We should add a commit/PR addressing all occurences

File: src/backend/sgx/thread.rs
Comment (haraldh) [reply src/backend/sgx/thread.rs:46]: added comment and issue https://github.com/enarx/enarx/issues/2200

File: src/backend/sgx/thread.rs
Comment (haraldh) [reply src/backend/sgx/thread.rs:120]: in fact it is `sallyport::NULL`

Comment (haraldh) [reply src/backend/sgx/thread.rs]: pushed a fixup

File: crates/sallyport/src/guest/call/enarxcall/types/result.rs
Comment (haraldh) [reply crates/sallyport/src/guest/call/enarxcall/types/result.rs]: pushed a fixup

File: src/backend/sgx/thread.rs
Comment (haraldh) [reply src/backend/sgx/thread.rs:49]: pushed a fixup

**Comment by bstrie** [comment]
LGTM, but I won't approve so that Roman and Nathaniel can have a chance to review as well.

**Comment by haraldh** [comment]
rebased and squashed

**Review by rvolosatovs — COMMENTED** [review]
Overall LGTM, waiting for @npmccallum to take a look

**Review by npmccallum — DISMISSED** [review]

**Comment by npmccallum** [comment]
@haraldh I'm okay with the overall direction. Also, I sign off on data leakage (there is none). Mostly I had some questions and a few small changes.

File: crates/shim-sgx/src/handler/mod.rs
Comment (haraldh) [reply crates/shim-sgx/src/handler/mod.rs:208]: yes

File: crates/shim-sgx/src/handler/mod.rs
Comment (haraldh) [reply crates/shim-sgx/src/handler/mod.rs:202]: yes

File: crates/shim-sgx/src/thread.rs
Comment (haraldh) [reply crates/shim-sgx/src/thread.rs:106]: yes

**Comment by npmccallum** [comment]
@haraldh I dismissed my review so that I'm not blocking the merge.

File: crates/sallyport/src/guest/handler.rs
Comment (haraldh) [reply crates/sallyport/src/guest/handler.rs:1192]: kind of... if it differs from the expected, the thread should wake up and check, if it still wants to park and re-enter then.

File: crates/shim-sgx/src/handler/mod.rs
Comment (haraldh) [reply crates/shim-sgx/src/handler/mod.rs:174]: yeah, I'll return ENOTSUP

File: crates/shim-sgx/src/thread.rs
Comment (haraldh) [reply crates/shim-sgx/src/thread.rs:65]: well spotted, thanks

File: src/backend/parking.rs
Comment (haraldh) [reply src/backend/parking.rs:43]: changed

File: src/backend/parking.rs
Comment (haraldh) [reply src/backend/parking.rs:55]: If the value is different, there is no need to do the `futex`, so we can return immediately.
Also, this is _the_ return point of the loop. `futex` can return without the value changed, so we have to check anyway.

File: crates/shim-sgx/src/handler/mod.rs
Comment (haraldh) [reply crates/shim-sgx/src/handler/mod.rs:200]: added

File: crates/sallyport/src/guest/call/enarxcall/park.rs
Comment (haraldh) [reply crates/sallyport/src/guest/call/enarxcall/park.rs:14]: `expected_val` ?

File: crates/sallyport/src/guest/handler.rs
Comment (haraldh) [reply crates/sallyport/src/guest/handler.rs:1192]: changed to:
```rust
    /// Park the current thread
    ///
    /// # Arguments
    /// expected_val: park the thread, as long as the global parking state has this value
    /// timeout: the CLOCK_MONOTONIC time, when to timeout the park operation
    ///
    /// # Returns
    /// the actual value of the global parking state
```

File: crates/shim-sgx/src/handler/mod.rs
Comment (haraldh) [reply crates/shim-sgx/src/handler/mod.rs]: returned ENOTSUP

File: crates/shim-sgx/src/thread.rs
Comment (haraldh) [reply crates/shim-sgx/src/thread.rs:130]: done

File: crates/sallyport/src/guest/call/enarxcall/park.rs
Comment (rvolosatovs) [inline crates/sallyport/src/guest/call/enarxcall/park.rs:14]: nit; `expected` would probably suffice and be consistent with the shim implementations. This is good enough though

**Review by rvolosatovs — APPROVED** [review]

**Pull request merged**
